src/Microsoft.ML.Core/Utilities/ResourceManagerUtils.cs

```diff
@@ -110,80 +108,64 @@ public async Task<ResourceDownloadResults> EnsureResourceAsync(IHostEnvironment
+                await DownloadFromUrlWithRetryAsync(absoluteUrl.AbsoluteUri, fileName, timeout, filePath), absoluteUrl.AbsoluteUri);
         }
 
-        private async Task<string> DownloadFromUrlWithRetryAsync(IHostEnvironment env, IChannel ch, string url, string fileName,
-            int timeout, string filePath, int retryTimes = 5)
+        private async Task<string> DownloadFromUrlWithRetryAsync(string url, string fileName, int timeout, string filePath, int retryTimes = 5)
         {
             var downloadResult = "";
 
             for (int i = 0; i < retryTimes; ++i)
             {
-                var thisDownloadResult = await DownloadFromUrlAsync(env, ch, url, fileName, timeout, filePath);
+                var thisDownloadResult = await DownloadFromUrlAsync(url, fileName, timeout, filePath);
 
                 if (string.IsNullOrEmpty(thisDownloadResult))
                     return thisDownloadResult;
                 else
+                {
                     downloadResult += thisDownloadResult + @"\n";
-
+                    // Do not retry if the URL does not exist
+                    if (thisDownloadResult.Contains("does not exist"))
+                        return downloadResult;
+                }
                 await Task.Delay(10 * 1000);
             }
 
             return downloadResult;
         }
 
         /// <returns>Returns the error message if an error occurred, null if download was successful.</returns>
-        private async Task<string> DownloadFromUrlAsync(IHostEnvironment env, IChannel ch, string url, string fileName, int timeout, string filePath)
+        private async Task<string> DownloadFromUrlAsync(string url, string fileName, int timeout, string filePath)
         {
-            using (var webClient = new WebClient())
+            using (var webClient = new WebClientResponseUri())
             using (var downloadCancel = new CancellationTokenSource())
             {
                 bool deleteNeeded = false;
                 EventHandler disposed =
                     (object sender, EventArgs e) =>
                     {
                         if (File.Exists(filePath) && deleteNeeded)
-                            TryDelete(ch, filePath);
+                            TryDelete(filePath);
                     };
 
                 webClient.Disposed += disposed;
-                var t = Task.Run(() => DownloadResource(env, ch, webClient, new Uri(url), filePath, fileName, downloadCancel.Token));
+                var t = Task.Run(() => DownloadResource(webClient, new Uri(url), filePath, fileName, downloadCancel.Token));
 
                 UpdateTimeout(ref timeout);
                 var timeoutTask = Task.Delay(timeout).ContinueWith(task => default(Exception), TaskScheduler.Default);
-                ch.Info($"Downloading {fileName} from {url} to {filePath}");
+                Console.WriteLine($"Downloading {fileName} from {url} to {filePath}");
```

Comment:
Libraries like ML.NET should never be writing to the Console directly.

Reply:
Correct. 

This originated from my .NET fiddle (https://dotnetfiddle.net/5zs4c2) where I didn't have an `IChannel` implemented; hence wrote to Console. I'd recommend using just the updates in my original [comment](https://github.com/dotnet/machinelearning/pull/4943#discussion_r392572981). No need to remove `IChannel` and `IHostEnvironment`.

For the fiddle, I also had to use `using (var fh = env.CreateOutputFile(tempPath)) using (var ws = fh.CreateWriteStream())` instead of the existing `using (var fs = File.Create(tempPath)) using (var ws = new StreamWriter(fs))`, for the same reason. We should keep the older method.

Reply:
Got it, I reverted these changes in my latest commit.